Pyproject v1

MehmedGIT opened the pull request.

kba commented on pyproject.toml:
```diff
@@ -1,3 +1,13 @@
+"Homepage" = "https://github.com/OCR-D/ocrd-webapi-implementation"
+
+[tool.setuptools]
+packages = ["ocrd_webapi"]
```
I'm curious: Why is that information (`packages`) in both `pyproject.toml` and `setup.cfg`?

Also, can you do `pip install -e .` without a setup.py?

kba commented on setup.cfg:
```diff
@@ -0,0 +1,77 @@
+packages =
+	ocrd_webapi
+	ocrd_webapi.routers
+install_requires =
```
Is this list meant to be exhaustive (including transitive dependencies)? And can you specify version ranges or omit the explicit versions?

joschrew replied:
What I am doing with the versions is just that: I write the code and if I need a library than I install it in the venv via pip. And to write these `install_requires` or requirements.txt I just execute `pip freeze` and put this in requirements.txt or `install_requires`. 
To me it seems your questions assume that there is a deeper sense (or good understanding)  behind what I do here but I have to admit I am mostly just stupidly trying until it works.

kba replied:
Haha, no need to be self-critical, `pip freeze` is exactly the right approach in most cases, it guarantees that the set of dependencies work together. With setup.py it was often argued, to generate the `requirements.txt` by doing `pip freeze` and manually edit the `install_requires` in setup.py for less restrictive dependencies. 

So, `pip freeze` is a good start. Then you can get an understanding of the inter-dependencies of packages with `pipdeptree`. Then I would prune all the transitive dependencies, e.g. ocrd brings in Pillow, bagit etc. And finally, unless there is a very strong reason for `==` version pinning, either omit the version or use `>=`.

Nothing pressing, this can be adapted later. I was only curious.

joschrew commented:
I don't know why a pull request was created here, I think it is automatically because we merged a branch.
At first I used pyproject.toml just for the tests to set env's and some stuff and for flake8. Later I wanted to use it for packaging, but realized this does not work with python3.6 so I created a setup.cfg. I don't know if it works with or without pyproject.toml but I keep that for tests.
I don't really understand how this should work I just created an example project on my machine and modified the setup.cfg from ocrd-webapi-implementation until I was able to use the ocrd-webapi-implementation as a library. So there is no specific intention behind anything i did here just to make it work to import it as a library.

kba commented:
Gotcha, as I said, I was only curious because once we leave Python 3.6 behind, we can change the setup to use `pytproject.toml` in most projects. But that is not time-essential.

> I don't know why a pull request was created here, I think it is automatically because we merged a branch.

Always a good idea to create PR to have discussions like this.

MehmedGIT commented:
In order to merge from the website interface I had to create a PR. That's the reason for the PR.

I have pushed a `setup.py` and removed the `setup.cfg` and `pyproject.toml` in the `refactoring` branch. I also cleaned the `requirements_3.7.txt`. Seems I am able to reuse the routers in Operandi without issues. The only problem is the DB which i have to start manually but this is easily fixable once we clearly separate the local install from docker. Ideally, the local install should not depend on anything based on docker. The dockerized version of the WebAPI can then just reuse the locally available configurations and source code and builded over the ocrd-all image.

joschrew commented:
My thoughts:
You focus on python3.7 but currently I need python3.6 to run the webapi standalone because it has to extend ocrd-all. Maybe I should change that because in the near future ocrd-all should run with python3.7 as well. And for the meantime I could try to build a ocrd-all with python3.7 by myself.
I was not able to use pyproject.toml with python3.6, but I think it should works with python3.7. I think we should try that because I think pyproject.toml is the way to go nowadays. Additionally the tests currently need he pyproject.toml because the variables are set there.
